### PR TITLE
Search functionality

### PR DESCRIPTION
**Pull request by keithudock**
- User can search for text using plain text or regular expressions (matches full string only)
- User can specify whether they want to create or delete highlight on search
- User can select Regex flags via a dropdown menu
- console logging
  - Logs word being searched for
  - Logs position of word if found
  - Logs flags selected in dropdown
  - Logs word was added to annotations list

### CODE REVIEW

**Comment by ghost** [comment]
@keithudock Could you merge these conflicts?

**Comment by keithudock** [comment]
conflicts fixed

**Comment by tmarkn** [comment]
Search delete seems to be broken

**Comment by keithudock** [comment]
How so?

**Comment by tmarkn** [comment]
- It doesn't seem to delete anything. My console says "Removing annotation: 'undefined'".
- Also, searching nothing highlights everything.
- It's also highlighting globally regardless of the flag.
- Edit: and searching without a label should be prevented

This is all in plain text.

**Comment by keithudock** [comment]
Hmmm I’ll check it in the morning. Can you dm me some test cases in discord?

**Comment by keithudock** [comment]
@tmarkn bugs fixed in latest commit

**Review by ghost — CHANGES_REQUESTED** [review]
@keithudock Some things I noticed:

- You can repeatedly add the same annotation to the list by searching again
- Log output when option set to Search to Delete: Removing annotation: 'undefined' from [246tahosh]
- Is there an option to disable/enable regex? I feel having both regular ctrl+f functionality and regex would be clunky

**Comment by tmarkn** [comment]
> * You can repeatedly add the same annotation to the list by searching again

We can ignore this for now. I was going to change the addAnnotation functionality to prevent this.

> * Is there an option to disable/enable regex? I feel having both regular ctrl+f functionality and regex would be clunky

The way it searches is using regex. We would replace the string with escape characters [(see here)](https://stackoverflow.com/questions/3561493/is-there-a-regexp-escape-function-in-javascript) and have a toggle for it.

File: static/js/search.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                tagModel.currentDoc.annotations.forEach( function(annotation){
+                    if(annotation.content.match(regex)){
+                        console.log("Found " + annotation.content + "\nStart postion: " + position + "\nEnd Position: " + (position + annotation.content.length));
+                        tagModel.removeAnnotation(annotation);
+                    }
+                    position += annotation.content.length+1; // the plus 1 accounts for the space at the end
+                })
+                renderTextareaHighlights();
+            }
+        }
```
Comment (tmarkn) [inline static/js/search.js]: Duplicate code.
Move to a function and call the function instead.

Comment (keithudock) [reply static/js/search.js]: stopped having a stroke and fixed it in the latest commit

**Comment by ghost** [comment]
@keithudock Annotation lists arent being updated for me (when searching)

**Comment by tmarkn** [comment]
> @keithudock Annotation lists arent being updated for me (when searching)

I think he started the branch before I fixed the annotations list.

**Comment by keithudock** [comment]
what tony said

**Review by ghost — DISMISSED** [review]

**Review by tuj52136 — DISMISSED** [review]
serchin

**Review by tuj52136 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**